### PR TITLE
Implements covariation structures as defined by Aitchison (1986) in definitions 4.4, 4.5 and 4.6

### PR DESCRIPTION
**Pull request by mrr00b00t**
This is a clean PR. As discussed earlier, the other one was poluted. I would appreciate if you could review this piece of code.

### CODE REVIEW

**Comment by mrr00b00t** [comment]
> 
> 
> This is a clean PR. As discussed earlier, the other one was poluted. I would appreciate if you could review this piece of code.

Furthermore, I have the test code for these new features, but it is dependent on the approval of this PR.

File: src/CoDa.jl [hunk 1]
```diff
@@ -9,6 +9,7 @@ using Tables
 using DataFrames
 using StatsBase
 using StaticArrays
+using Statistics
 using UnicodePlots
 using LinearAlgebra
 using FillArrays
```
Comment (juliohm) [inline src/CoDa.jl]:
```suggestion
using UnicodePlots
using LinearAlgebra
using FillArrays
using Statistics
```

File: src/CoDa.jl [hunk 2]
```diff
@@ -18,11 +19,19 @@ import Base: names, getproperty
...
 include("utils.jl")
 
 export
+  # covariances
+  designmatrix,
+  compvarmatrix,
+  variationmatrix,
+  lrcovmatrix,
+  clrcovmatrix,
+
```
Comment (juliohm) [inline src/CoDa.jl]: Please move these exports to before the matrices exports. Try to follow the same order of the include statements.

File: src/covariances.jl [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENCE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    designmatrix(comps)
+    
+Converts a vector of Composition objects into a Matrix of numbers.
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
Converts a vector of `Composition{D}` objects into the N by D design matrix.
```

File: src/covariances.jl [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+"""
+function designmatrix(comps)
+    reduce(hcat, components.(comps))'
+end
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
designmatrix(comps) = reduce(hcat, components.(comps))'
```

File: src/covariances.jl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+"""
+    compvarmatrix(comps)
+
+Returns the compositional variation array, from definition 4.3 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function compvarmatrix(comps)
+    X = designmatrix(comps)
```
Comment (juliohm) [inline src/covariances.jl]: Incorrect indentation, please use 2 spaces.

File: src/covariances.jl [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    N, D = size(X)
+    cva = zeros(D, D)
+
+    for i in 1:D
+        for j in i+1:D
+            log_ratios = log.(X[:, i] ./ X[:, j])
+            cva[j, i] = sum(log_ratios) / N
+            cva[i, j] = sum((log_ratios .- cva[j, i]).^2 ) / N
```
Comment (juliohm) [inline src/covariances.jl]: Please use `mean` and `var` from Statistics now that you have added it as a dependency.

File: src/covariances.jl [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+Returns the compositional variation array, from definition 4.3 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function compvarmatrix(comps)
+    X = designmatrix(comps)
+    N, D = size(X)
+    cva = zeros(D, D)
+
+    for i in 1:D
+        for j in i+1:D
+            log_ratios = log.(X[:, i] ./ X[:, j])
```
Comment (juliohm) [inline src/covariances.jl]: Rename this variable to `lr`

File: src/covariances.jl [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+"""
+    compvarmatrix(comps)
+
+Returns the compositional variation array, from definition 4.3 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function compvarmatrix(comps)
+    X = designmatrix(comps)
+    N, D = size(X)
+    cva = zeros(D, D)
```
Comment (juliohm) [inline src/covariances.jl]: Rename this variable to `A`, we use capital letters for matrices in general.

File: src/covariances.jl [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    for i in 1:D
+        for j in i+1:D
+            log_ratios = log.(X[:, i] ./ X[:, j])
+            cva[j, i] = sum(log_ratios) / N
+            cva[i, j] = sum((log_ratios .- cva[j, i]).^2 ) / N
+        end
+    end
+
+    return cva
```
Comment (juliohm) [inline src/covariances.jl]: Again, we don't use `return` in the code style.

File: src/covariances.jl [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+"""
+    lrcovmatrix(comps)
+
+Return the log ratio covariance matrix, definition 4.5 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function lrcovmatrix(comps)
+    lrcomps = alr.(comps)
+    lrmatrix = reduce(hcat, lrcomps)'
+
```
Comment (juliohm) [inline src/covariances.jl:44]:
```suggestion
```

File: src/covariances.jl [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+"""
+    clrcovmatrix(comps)
+
+Return the centered log ratio covariance matrix, definition 4.6 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function clrcovmatrix(comps)
+    clrcomps = clr.(comps)
+    clrmatrix = reduce(hcat, clrcomps)'
+
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
```

File: src/covariances.jl [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    clrcovmatrix(comps)
+
+Return the centered log ratio covariance matrix, definition 4.6 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function clrcovmatrix(comps)
+    clrcomps = clr.(comps)
+    clrmatrix = reduce(hcat, clrcomps)'
+    
+    Γ = cov(clrmatrix, corrected=false)
+end
```
Comment (juliohm) [inline src/covariances.jl:76]:
```suggestion
end
```

Comment (juliohm) [reply src/covariances.jl:76]: Please configure your editor EOL character so that it doesn't have anything specific to Windows.

**Review by juliohm — CHANGES_REQUESTED** [review]
Thank you @mrr00b00t for the updated PR. Please fix the code style and naming. After that we need tests.

**Comment by juliohm** [comment]
@mrr00b00t I have accepted the suggestions directly on GitHub. In order to continue working on this branch, please pull the changes first.

File: src/covariances.jl (hunk 9 above)
Comment (mrr00b00t) [reply src/covariances.jl:30]: When I remove the return this function doesn't return anything.

File: src/covariances.jl [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    designmatrix(comps)
+    
+Converts a vector of `Composition{D}` objects into the N by D design matrix.
+"""
+designmatrix(comps) = reduce(hcat, components.(comps))'
+
+"""
+    compvarmatrix(comps)
+
+Returns the compositional variation array, from definition 4.3 of Aitchson - The Statistical Analysis of Compositional Data.
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
Returns the variation array `A` such that:

- `A[i,j] = E[log(x[i]/x[j])]` for `i > j`
- `A[i,j] = Var(log(x[i]/x[j]))` for `i < j`
- `A[i,j] = 0` for `i = j`
```

File: src/covariances.jl [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+function compvarmatrix(comps)
+  X = designmatrix(comps)
+  N, D = size(X)
+  Α = zeros(D, D)
+
+  for i in 1:D
+    for j in i+1:D
+      lr = log.(X[:, i] ./ X[:, j])
+      Α[j, i] = mean(lr)
+      Α[i, j] = var(lr, corrected=false)
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
      Α[j,i] = mean(lr)
      Α[i,j] = var(lr, corrected=false)
```

File: src/covariances.jl [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+Returns the compositional variation array, from definition 4.3 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function compvarmatrix(comps)
+  X = designmatrix(comps)
+  N, D = size(X)
+  Α = zeros(D, D)
+
+  for i in 1:D
+    for j in i+1:D
+      lr = log.(X[:, i] ./ X[:, j])
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
      lr = log.(X[:,i] ./ X[:,j])
```

File: src/CoDa.jl [hunk 16]
```diff
@@ -29,6 +31,13 @@ export
   norm, dot, ⋅,
   distance,
 
+  # covariances
+  designmatrix,
+  compvarmatrix,
```
Comment (juliohm) [inline src/CoDa.jl]: Maybe we should rename this to `lrmeanvar`.

File: src/covariances.jl [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+Return the variation matrix, definition 4.4 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function variationmatrix(comps)
+  X = designmatrix(comps)
+  N, D = size(X)
+  Τ = zeros(D, D)
+
+  for i in 1:D
+    for j in 1:D
+      Τ[i, j] = var(log.(X[:, i] ./ X[:, j]), corrected=false)
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
      Τ[i,j] = var(log.(X[:,i] ./ X[:,j]), corrected=false)
```

File: src/CoDa.jl (hunk 16 above)
Comment (juliohm) [reply src/CoDa.jl]: I think `lrarray` would be even better.

File: src/covariances.jl [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    end
+  end
+
+  Τ
+end
+
+"""
+    lrcovmatrix(comps)
+
+Return the log ratio covariance matrix, definition 4.5 of Aitchson - The Statistical Analysis of Compositional Data.
```
Comment (juliohm) [inline src/covariances.jl]: Please write down the definitions in the docstring like I did with the variation array. Users don't have the book at hand to consult the definition.

File: src/CoDa.jl (hunk 16 above)
Comment (juliohm) [reply src/CoDa.jl]: Actually let's call it `variationarray`

File: src/covariances.jl [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+"""
+function lrcovmatrix(comps)
```
Comment (juliohm) [inline src/covariances.jl]: Also, maybe we should just rename to `alrcov` because matrix is implicit.

File: src/covariances.jl [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+  lrmatrix = reduce(hcat, lrcomps)'
+  Σ = cov(lrmatrix, corrected=false)
+end
+
+"""
+    clrcovmatrix(comps)
+
+Return the centered log ratio covariance matrix, definition 4.6 of Aitchson - The Statistical Analysis of Compositional Data.
+"""
+function clrcovmatrix(comps)
```
Comment (juliohm) [inline src/covariances.jl]: and this one to `clrcov`

File: src/CoDa.jl [hunk 21]
```diff
@@ -29,6 +31,13 @@ export
   norm, dot, ⋅,
   distance,
 
+  # covariances
+  designmatrix,
+  variationmatrix,
```
Comment (juliohm) [inline src/CoDa.jl]: Sorry for renaming requests again, but the picture is becoming clearer now. Can we rename this one to `variation` without the matrix suffix? I think it is also implicit.

File: src/CoDa.jl [hunk 22]
```diff
@@ -29,6 +31,13 @@ export
   norm, dot, ⋅,
   distance,
 
+  # covariances
+  designmatrix,
```
Comment (juliohm) [inline src/CoDa.jl]: Rename to `design`

File: src/covariances.jl [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+Returns the logratio covariance matrix `Σ` such that:
+
+- `Σ[i,j] = Cov(log(x[i]/x[D]), log(x[j]/x[D]))` for `i, j = 1, ..., d`
+"""
+function alrcov(comps)
+  lrcomps = alr.(comps)
+  lrmatrix = reduce(hcat, lrcomps)'
+  Σ = cov(lrmatrix)
+end
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
alrcov(comps) = reduce(hcat, alr.(comps))' |> cov
```

File: src/covariances.jl [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+Returns the centred logratio covariance matrix `Γ` such that:
+
+- `Γ[i,j] = Cov(log(x[i]/g(x)), log(x[j]/g(x)))` for `i, j = 1, ..., D`,
+where g(x) is the geometric mean.
+"""
+function clrcov(comps)
+  clrcomps = clr.(comps)
+  clrmatrix = reduce(hcat, clrcomps)'
+  Γ = cov(clrmatrix)
+end
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
clrcov(comps) = reduce(hcat, clr.(comps))' |> cov
```

File: src/CoDa.jl (hunk 22 above)
Comment (mrr00b00t) [reply src/CoDa.jl]: Solved in commit 36ccc55f38c7eadce190fb1f9eb9b0c780a5ca47.

File: src/CoDa.jl (hunk 21 above)
Comment (mrr00b00t) [reply src/CoDa.jl]: Solved in commit 36ccc55f38c7eadce190fb1f9eb9b0c780a5ca47.

File: src/covariances.jl [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+
+"""
+    alrcov(comps)
+
+Returns the logratio covariance matrix `Σ` such that:
+
+- `Σ[i,j] = Cov(log(x[i]/x[D]), log(x[j]/x[D]))` for `i, j = 1, ..., d`
+"""
+alrcov(comps) = reduce(hcat, alr.(comps))' |> cov
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
alrcov(comps) = cov(reduce(hcat, alr.(comps)), dims=2)
```

File: src/covariances.jl [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+
+"""
+    clrcov(comps)
+
+Returns the centred logratio covariance matrix `Γ` such that:
+
+- `Γ[i,j] = Cov(log(x[i]/g(x)), log(x[j]/g(x)))` for `i, j = 1, ..., D`,
+where g(x) is the geometric mean.
+"""
+clrcov(comps) = reduce(hcat, clr.(comps))' |> cov
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
clrcov(comps) = cov(reduce(hcat, clr.(comps)), dims=2)
```

File: src/CoDa.jl [hunk 27]
```diff
@@ -29,6 +31,13 @@ export
   norm, dot, ⋅,
   distance,
 
+  # covariances
+  design,
+  variation,
+  alrcov,
+  clrcov,
+  variationarray,
```
Comment (juliohm) [inline src/CoDa.jl]:
```suggestion
  lrarray,
```

File: src/covariances.jl [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+"""
+    variationarray(comps)
+
+Returns the variation array `A` such that:
+
+- `A[i,j] = E[log(x[i]/x[j])]` for `i > j`
+- `A[i,j] = Var(log(x[i]/x[j]))` for `i < j`
+- `A[i,j] = 0` for `i = j`
+"""
+function variationarray(comps)
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
function lrarray(comps)
```

File: src/covariances.jl [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+
+Returns the centred logratio covariance matrix `Γ` such that:
+
+- `Γ[i,j] = Cov(log(x[i]/g(x)), log(x[j]/g(x)))` for `i, j = 1, ..., D`,
+where g(x) is the geometric mean.
+"""
+clrcov(comps) = cov(reduce(hcat, clr.(comps)), dims=2)
+
+"""
+    variationarray(comps)
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
    lrarray(comps)
```

File: test/covariances.jl [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+  @test Τ ≈ J(D) * Diagonal(Γ₀) + Diagonal(Γ₀) * J(D) - 2 * Γ₀
+
+  ## Test matrix transformation from Σ to Γ (4.28)
+  @test Γ ≈ F(d)' * inv(H(d)) * Σ * inv(H(d)) * F(d)
+
+  ## Test matrix transformation from Γ to Τ (4.29)
+  @test Τ ≈ J(D) * Diagonal(Γ) + Diagonal(Γ) * J(D) - 2 * Γ
+
+  ## Test matrix transformation from Γ to Σ (4.30)
+  @test Σ ≈ F(d) * Γ * F(d)'
```
Comment (juliohm) [inline test/covariances.jl:28]: Very cool test set! 💯

File: src/covariances.jl [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+    variation(comps)
+
+Returns the variation matrix `Τ` such that:
+
+- `Τ[i,j] = Var(log(x[i]/x[j]))` for `i, j = 1, ..., D`
+"""
+function variation(comps)
+  X = design(comps)
+  N, D = size(X)
+  Τ = zeros(D, D)
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion
  Τ = Matrix{Float64}(undef, D, D)
```

File: src/covariances.jl [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+
+Returns the variation matrix `Τ` such that:
+
+- `Τ[i,j] = Var(log(x[i]/x[j]))` for `i, j = 1, ..., D`
+"""
+function variation(comps)
+  X = design(comps)
+  N, D = size(X)
+  Τ = Matrix{Float64}(undef, D, D)
+
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion

  Τ = Matrix{Float64}(undef, D, D)
```

File: src/covariances.jl [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+- `A[i,j] = E[log(x[i]/x[j])]` for `i > j`
+- `A[i,j] = Var(log(x[i]/x[j]))` for `i < j`
+- `A[i,j] = 0` for `i = j`
+"""
+function lrarray(comps)
+  X = design(comps)
+  N, D = size(X)
+  Α = zeros(D, D)
+
```
Comment (juliohm) [inline src/covariances.jl]:
```suggestion

  Α = Matrix{Float64}(undef, D, D)
```

File: src/covariances.jl [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+  X = design(comps)
+  N, D = size(X)
+
+  Α = Matrix{Float64}(undef, D, D)
+  for i in 1:D
+    for j in i+1:D
+      lr = log.(X[:,i] ./ X[:,j])
+      Α[j,i] = mean(lr)
+      Α[i,j] = var(lr)
+    end
```
Comment (juliohm) [inline src/covariances.jl:71]:
```suggestion
    end
    A[i,i] = 0.0
```

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
Amazing PR! 🎉 ❤️

**Pull request merged**